src/server/typingsInstaller/typingsInstaller.ts

```diff
@@ -367,51 +385,112 @@ namespace ts.server.typingsInstaller {
             }
         }
 
-        private watchFiles(projectName: string, files: string[]) {
+        private watchFiles(projectName: string, files: string[], projectRootPath: Path) {
             if (!files.length) {
                 // shut down existing watchers
                 this.closeWatchers(projectName);
                 return;
             }
 
             let watchers = this.projectWatchers.get(projectName);
+            const toRemove = createMap<FileWatcher>();
             if (!watchers) {
                 watchers = createMap();
                 this.projectWatchers.set(projectName, watchers);
             }
+            else {
+                copyEntries(watchers, toRemove);
+            }
 
-            watchers.isInvoked = false;
             // handler should be invoked once for the entire set of files since it will trigger full rediscovery of typings
+            watchers.isInvoked = false;
+
             const isLoggingEnabled = this.log.isEnabled();
-            mutateMap(
-                watchers,
-                arrayToSet(files),
-                {
-                    // Watch the missing files
-                    createNewValue: file => {
-                        if (isLoggingEnabled) {
-                            this.log.writeLine(`FileWatcher:: Added:: WatchInfo: ${file}`);
-                        }
-                        const watcher = this.installTypingHost.watchFile(file, (f, eventKind) => {
-                            if (isLoggingEnabled) {
-                                this.log.writeLine(`FileWatcher:: Triggered with ${f} eventKind: ${FileWatcherEventKind[eventKind]}:: WatchInfo: ${file}:: handler is already invoked '${watchers.isInvoked}'`);
-                            }
-                            if (!watchers.isInvoked) {
-                                watchers.isInvoked = true;
-                                this.sendResponse({ projectName, kind: ActionInvalidate });
-                            }
-                        }, /*pollingInterval*/ 2000);
-                        return isLoggingEnabled ? {
-                            close: () => {
-                                this.log.writeLine(`FileWatcher:: Closed:: WatchInfo: ${file}`);
-                            }
-                        } : watcher;
-                    },
-                    // Files that are no longer missing (e.g. because they are no longer required)
-                    // should no longer be watched.
-                    onDeleteValue: closeFileWatcher
+            const createProjectWatcher = (path: string, createWatch: (path: string) => FileWatcher) => {
+                toRemove.delete(path);
+                if (watchers.has(path)) {
+                    return;
+                }
+
+                watchers.set(path, createWatch(path));
+            };
+            const createProjectFileWatcher = (file: string): FileWatcher => {

```

Comment:
Is there a reason to prefer lambdas over functions or vice versa?

Reply:
lambdas are needed because we use ```this```